hazardous/metrics/_concordance_index.py

```diff
@@ -0,0 +1,439 @@
+    )
+    if pair_type == "b":
+        # For b type statistics, we only compare individuals who experienced the event
+        # of interest with individuals who experienced a competing event.
+        y_pred, duration, ipcw = (
+            y_pred[~mask_event],
+            duration[~mask_event],
+            ipcw[~mask_event],
+        )
+
+    stats = Counter()
+    for y_pred_i, duration_i, ipcw_i in zip(y_pred_event, duration_event, ipcw_event):
+        idx_acceptable, n_ties_times = _get_idx_acceptable(
+            event, duration, duration_i, pair_type
+        )
+        stats["n_ties_times"] += n_ties_times
+
+        stats["n_pairs"] += duration.shape[0] - idx_acceptable
+        stats["weighted_pairs"] += _compute_weights(
+            ipcw_i, ipcw[idx_acceptable:], pair_type
+        )
+
+        mask_ties_pred = np.abs(y_pred_i - y_pred[idx_acceptable:]) <= tied_tol
+        stats["n_ties_pred"] += mask_ties_pred.sum()
+        stats["weighted_ties_pred"] += _compute_weights(
+            ipcw_i, ipcw[idx_acceptable:][mask_ties_pred], pair_type
+        )
+
+        mask_concordant = y_pred_i > y_pred[idx_acceptable:]
+        stats["n_concordant_pairs"] += (mask_concordant & ~mask_ties_pred).sum()
+        stats["weighted_concordant_pairs"] += _compute_weights(
+            ipcw_i, ipcw[idx_acceptable:][mask_concordant & ~mask_ties_pred], pair_type
+        )
+
+    return stats
+
+
+def _sort_by_duration(event, duration, y_pred, ipcw, pair_type):
+    """Sort the predictions and duration by the event duration.
+
+    The pair type selects whether we sort duration by ascending or descending order.
+    Indeed, to be comparable:
+    - A pair of type A requires (T_i < T_j) | ((T_i = T_j) & (D_j = 0))
+    - A pair of type B requires T_i >= T_j
+    """
+    if pair_type == "b":
+        duration *= -1
+    # Sort by ascending duration first, then by descending event.
+    # After reordering, we would have:
+    # duration = [10, 10, 10, 11]
+    # event = [2, 1, 0, 1]
+    indices = np.lexsort((-event, duration))
+    event = event[indices]
+    duration = duration[indices]
+    y_pred = y_pred[indices]
+    ipcw = ipcw[indices]
+
+    return event, duration, y_pred, ipcw
+
+
+def _get_idx_acceptable(event, duration, duration_i, pair_type):
+    """Returns idx_acceptable and n_times_ties"""
+    if pair_type == "a":
+        # We select all acceptable pairs by only keeping elements strictly higher than
+        # `duration`, which corresponds to A_{ij} = I(T_i < T_j).
+        # We also select censored pairs where T_i = T_j and D_j = 0.
+        # Using searchsorted with `side=right` returns the index of the smallest
+        # element in `duration` strictly higher than `duration_i`.
+        idx_acceptable = np.searchsorted(duration, duration_i, side="right")
+        idx_with_ties = np.searchsorted(duration, duration_i, side="left")
+        n_censored_ties = (event[idx_with_ties:idx_acceptable] == 0).sum()
```

Comment:
I have to check, but I think you also count the competing events in that case.

Reply:
I need to revamp this slightly, in the meantime I addressed your doc remarks

Reply:
Great ! Do not hesitate to ping me when the revamp is finalized.